src/java.base/share/classes/java/lang/Class.java

```diff
@@ -1832,12 +1869,29 @@ private String getCanonicalName0() {
         }
     }
 
+    /**
+     * {@return {@code true} if and only if the underlying class
+     * is an unnamed class}
+     *
+     * @apiNote
+     * An unnamed class is not an {@linkplain #isAnonymousClass anonymous class}.
+     *
+     * @since 21
+     */
+    @PreviewFeature(feature=PreviewFeature.Feature.UNNAMED_CLASSES,
+                    reflective=true)
+    public boolean isUnnamedClass() {
+        return isSynthetic() && PreviewFeatures.isEnabled() && isTopLevelClass();
```

Comment:
Should this include more tests such as not abstract, final, package access, in an unnamed package, etc to accurately determine it's an unnamed class?

Reply:
Hmm. I think adding at least a check for ACC_FINAL should be fine as it should have near-zero cost on top of the existing predicate once the method get compiled. (In future iterations of the JEP, the idiom to identify isUnnamed will likely change.)

Reply:
As mentioned in https://bugs.openjdk.org/browse/JDK-8308916 the expression should be `PreviewFeatures.isEnabled() && isSynthetic() &&  isTopLevelClass()` as `PreviewFeatures.isEnabled()` is a private static final test that will effectively make the full expression constant `false` when not in preview mode.

Reply:
Checking package access should also have near-zero cost.

Reply:
Will this accidentally return `true` for classes from other JVM programming languages?